File: app/src/test/java/de/westnordost/osmagent/meta/AbbreviationsTest.java
package de.westnordost.osmagent.meta;

import junit.framework.TestCase;

import java.io.ByteArrayInputStream;
import java.io.UnsupportedEncodingException;
import java.util.Locale;

public class AbbreviationsTest extends TestCase
{
	public void testCapitalizeFirstLetter()
	{
		assertEquals("Straße", abbr("str: straße").getExpansion("str", true, true));
	}

	public void testWithAbbrDot()
	{
		assertEquals("Straße", abbr("str: straße").getExpansion("str.", true, true));
	}

	public void testIgnoreCase()
	{
		assertEquals("Straße", abbr("sTr: Straße").getExpansion("StR", true, true));
	}

	public void testExpectOwnWordByDefault()
	{
		assertEquals(null, abbr("st: street").getExpansion("Hanswurst", true, true));
	}

	public void testConcatenable()
	{
		assertEquals("Konigstraat", abbr("...str: Straat").getExpansion("Konigstr", true, true));
	}

	public void testConcatenableEnd()
	{
		assertEquals("Konigstraat", abbr("...str$: Straat").getExpansion("Konigstr", true, true));
	}

	public void testConcatenableWorksNormallyForNonConcatenation()
	{
		assertEquals("Straat", abbr("...str: Straat").getExpansion("str", true, true));
	}

	public void testOnlyFirstWord()
	{
		Abbreviations abbr = abbr("^st: Saint");
		assertEquals(null, abbr.getExpansion("st.", false, false));
		assertEquals("Saint", abbr.getExpansion("st.", true, false));
	}

	public void testOnlyLastWord()
	{
		Abbreviations abbr = abbr("str$: Straße");
		assertEquals(null, abbr.getExpansion("str", true, false));
		assertEquals(null, abbr.getExpansion("str", true, true));
		assertEquals("Straße", abbr.getExpansion("str", false, true));
	}

	public void testUnicode()
	{
		assertEquals("Блок", abbr("бл: Блок").getExpansion("бл", true, true));
	}

	public void testLocaleCase()
	{
		assertEquals("Блок", abbr("бл: блок").getExpansion("Бл", true, true));
	}

	public void testFindAbbreviation()
	{
		assertFalse(abbr("str: Straße").containsAbbreviations("stri stra straße"));
		assertTrue(abbr("str: Straße").containsAbbreviations("stri str straße"));
	}

	private Abbreviations abbr(String input)
	{
		try
		{
			return new Abbreviations(new ByteArrayInputStream(input.getBytes("utf-8")), Locale.US);
		}
		catch (UnsupportedEncodingException e)
		{
			throw new RuntimeException(e);
		}
	}
}

File: app/src/commonTest/kotlin/de/westnordost/streetcomplete/data/meta/AbbreviationsTest.kt
package de.westnordost.streetcomplete.data.meta

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class AbbreviationsTest {
    @Test fun `removes abbreviation dot`() {
        val abbr = Abbreviations(mapOf("str" to "straße"))
        assertEquals("straße", abbr.getExpansion("str.", true, true))
    }

    @Test fun `ignores case`() {
        val abbr = Abbreviations(mapOf("sTr" to "straße"))
        assertEquals("straße", abbr.getExpansion("StR", true, true))
    }

    @Test fun `expects own word by default`() {
        val abbr = Abbreviations(mapOf("st" to "street"))
        assertNull(abbr.getExpansion("Hanswurst", true, true))
    }

    @Test fun `concatenable expansion`() {
        val abbr = Abbreviations(mapOf("...str" to "Straat"))
        assertEquals("Konigstraat", abbr.getExpansion("Konigstr", true, true))
    }

    @Test fun `concatenable expansion on end`() {
        val abbr = Abbreviations(mapOf("...str$" to "Straat"))
        assertEquals("Konigstraat", abbr.getExpansion("Konigstr", true, true))
    }

    @Test fun `concatenable works normally for non-concatenation`() {
        val abbr = Abbreviations(mapOf("...str" to "straat"))
        assertEquals("straat", abbr.getExpansion("str", true, true))
    }

    @Test fun `get expansion of only first word`() {
        val abbr = Abbreviations(mapOf("^st" to "saint"))
        assertNull(abbr.getExpansion("st.", false, false))
        assertEquals("saint", abbr.getExpansion("st.", true, false))
    }

    @Test fun `get expansion of only last word`() {
        val abbr = Abbreviations(mapOf("str$" to "straße"))
        assertNull(abbr.getExpansion("str", true, false))
        assertNull(abbr.getExpansion("str", true, true))
        assertEquals("straße", abbr.getExpansion("str", false, true))
    }

    @Test fun `uses unicode`() {
        val abbr = Abbreviations(mapOf("бл" to "блок"))
        assertEquals("блок", abbr.getExpansion("бл", true, true))
    }

    @Test fun `finds abbreviation`() {
        val abbr = Abbreviations(mapOf("str" to "Straße"))
        assertFalse(abbr.containsAbbreviations("stri stra straße"))
        assertTrue(abbr.containsAbbreviations("stri str straße"))
    }
}
